Sensors 101 update (#283)
astronomer/airflow-guides: guides/what-is-a-sensor.md

In this guide, we'll cover the basics of using sensors in Airflow, best practices for implementing sensors in production, and new sensor-related features in Airflow 2 like smart sensors and deferrable operators. For more in-depth guidance on these topics, check out our [mastering sensors webinar](https://www.astronomer.io/events/webinars/creating-data-pipelines-using-master-sensors/).

## Sensor Basics

Sensors are conceptually simple: they are operators that check if a condition is satisfied on a particular interval. If the condition has been met, the task is marked successful and the DAG can move on to downstream tasks. If the condition hasn't been met, the sensor waits for another interval before checking again. 

All sensors inherit from the [`BaseSensorOperator`](https://github.com/apache/airflow/blob/main/airflow/sensors/base.py). and have the following parameters:

- `mode`: How the sensor operates. There are two options:
    - `poke`: This is the default mode. When using `poke`, the sensor occupies a worker slot for the entire execution time and sleeps between pokes. This mode is best if you expect a short runtime for the sensor.

>     - `reschedule`: When using this mode, the sensor will release its worker slot the criteria is not met and will reschedule the next check at a later time. This mode is best if you expect a long runtime for the sensor, because it is less resource intensive and frees up workers for other tasks.

Reviewer: ```suggestion
    - `reschedule`: When using this mode, if the criteria is not met then the sensor releases its worker slot and reschedules the next check at a later time. This mode is best if you expect a long runtime for the sensor, because it is less resource intensive and frees up workers for other tasks.
```

Reviewer: How do you feel about "condition" vs. "criteria" I kind of like "condition" because it implies some sort of true/false state, whereas a criteria seems like more of a prerequisite or standard that isn't as binary.

Author: Yeah I like that suggestion, good to use "condition" throughout

## Sensor Best Practices

Sensors are easy to implement, but there are a few things to keep in mind when using them to ensure you are getting the best possible Airflow experience. The following tips will help you avoid any performance issues when using sensors:

- **Always** define a meaningful `timeout` parameter for your sensor. The default for this parameter is 7 days, which is a long time for your sensor to be running! When you implement a sensor, take care to know your use case and how long you expect your sensor to be waiting and define your timeout accordingly.
- Whenever possible and especially for long-running sensors, use the `reschedule` mode so your sensor is not constantly occupying a worker slot. This will help avoid deadlocks in Airflow from sensors taking all the available worker slots. The exception to this rule is the following point:
- If your `poke_interval` is very short (less than about 5 minutes), use the `poke` mode. Using `reschedule` mode in this case can overload your scheduler.

> - Define a meaningful `poke_interval` based on your use case. There is no need for the task to check the condition every 30 seconds (the default) if you know the total amount of waiting time is likely to be 30 minutes.

Reviewer: ```suggestion
- Define a meaningful `poke_interval` based on your use case. There is no need for a task to check a condition every 30 seconds (the default) if you know the total amount of waiting time is likely to be 30 minutes.
```

## Smart Sensors

[Smart sensors](https://airflow.apache.org/docs/apache-airflow/stable/concepts/smart-sensors.html) are a relatively new feature released with Airflow 2.0, where sensors are executed in batches using a centralized process. This eliminates a major drawback of classic sensors, which use one process per sensor and therefore can consume considerable resources at scale for longer running tasks. With smart sensors, sensors in your DAGs are registered to the metadata database, and then a separate DAG fetches those sensors and manages processing them. 

![Smart Sensors](https://assets2.astronomer.io/main/guides/sensors-101/smart_sensors_architecture.png)

> Note that smart sensors are considered a beta feature, and may be changed in future versions of Airflow. They are somewhat complex to implement, and not very widely used to-date. They are also generally less preferable to deferrable operators (more on these in the section below), which are more flexible. For these reasons, we recommend using smart sensors only to advanced Airflow users who are running many sensors.

Reviewer: ```suggestion
Note that smart sensors are considered a beta feature, and so they might be changed in future versions of Airflow. They are somewhat complex to implement, and not very widely used to-date. They are also generally less preferable to deferrable operators (more on these in the section below), which are more flexible. For these reasons, we recommend using smart sensors only to advanced Airflow users who are running many sensors.
```

> To use smart sensors, you must be using Airflow 2+. You can then enable smart sensors using the following steps. 

Reviewer: ```suggestion
To use smart sensors, you must be using Airflow 2.0+. You can then enable smart sensors using the following steps:
```

2. Create your smart sensor class. Create a file in your `plugins/` directory and define a class for each type of sensor you want to make "smart". In this example, we implement smart sensors with the `FileSensor`, and the class looks like this:

    ```python
    from airflow.sensors.filesystem import FileSensor
    from airflow.utils.decorators import apply_defaults
    from typing import Any

    class SmartFileSensor(FileSensor):
        poke_context_fields = ('filepath', 'fs_conn_id') # <- Required

        @apply_defaults
        def __init__(self,  **kwargs: Any):
            super().__init__(**kwargs)

        def is_smart_sensor_compatible(self): # <- Required
            result = (
                not self.soft_fail
                and super().is_smart_sensor_compatible()
            )
            return result
    ```

    The class should inherit from the sensor class you are updating (e.g. `FileSensor`). It must include `poke_context_fields`, which specifies the arguments needed by your sensor, and the `is_smart_sensor_compatible` method, which tells Airflow this type of sensor is a smart sensor. Note when using smart sensors you cannot use soft fail or any callbacks.

>     Implementation of this class will vary depending on which sensor you are using, and some are more complex than others. For an example with the more complicated `ExternalTaskSensor`, see the [supporting repo](https://github.com/marclamberti/webinar-sensors).

Reviewer: ```suggestion
    The implementation of this class varies depending on which sensor you are using. For an example with the more complicated `ExternalTaskSensor`, see the [supporting repo](https://github.com/marclamberti/webinar-sensors).
```

> 3. Run Airflow or deploy the changes above to your Airflow deployment. Then turn on the `smart_sensor_group_shard_0` DAG to run your smart sensors. Note you may have more than one smart sensor DAG if you set your `SHARDS` parameter to greater than 1. 

Reviewer: ```suggestion
3. Deploy your code to Airflow and turn on the `smart_sensor_group_shard_0` DAG to run your smart sensors. Note you might have more than one smart sensor DAG if you set your `SHARDS` parameter to greater than 1. 
```

## Deferrable Operators

[Deferrable operators](https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html) (sometimes referred to as asynchronous operators) were released with Airflow 2.2 and were designed to eliminate the problem of *any* operator or sensor taking up a full worker slot for the entire time they are running. In other words, they solve the same problems as smart sensors, but for a much wider range of use cases. 

> For DAG authors, using built-in deferrable operators is no different from using regular operators (i.e. using them is much simpler than implementing smart sensors). You just need to ensure you have a `triggerer` process running in addition to your Scheduler. Currently, the `DateTimeSensorAsync` and `TimeDeltaSensorAsync` sensors are available out of the box in OSS Airflow. More deferrable operators will likely be released in future Airflow versions. For more on writing your own deferrable operators, check out the [Apache Airflow documentation](https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html#smart-sensors).

Reviewer: Do we want to mention anything about importing? I know Andrew + Jake wanted us to explain how to import deferrable operators `as` their non deferrable class names in our Cloud docs, but that advice could apply to the greater Airflow community as well.
